### PR TITLE
fix: fix hashmap thread unsafe

### PR DESCRIPTION
**Pull request by imp2002**
Fix: https://github.com/casbin/jcasbin/issues/292

I reproduced the bug mentioned in the issue. But not sure the exact reason, the bug appeared after this pr https://github.com/casbin/jcasbin/pull/265, I compared the code and found it caused by hashmap, then made this modification and it did work.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@shijianjs @tangyang9464 plz review

File: src/main/java/org/casbin/jcasbin/util/BuiltInFunctions.java [hunk 1]
```diff
@@ -334,26 +332,26 @@ public static boolean allMatch(String key1, String key2) {
     }
 
 
-    public static class GenerateGFunctionClass{
+    public static class GenerateGFunctionClass {
         // key:name such as g,g2  value:user-role mapping
-        private static Map<String, Map<String, AviatorBoolean>> memorizedMap = new HashMap<>();
+        private static Map<String, Map<String, AviatorBoolean>> memorizedMap = Collections.synchronizedMap(new HashMap<>());
```
Comment (tangyang9464) [inline src/main/java/org/casbin/jcasbin/util/BuiltInFunctions.java:337]: use ConcurrentHashMap

**Review by tangyang9464 — CHANGES_REQUESTED** [review]

**Review by tangyang9464 — APPROVED** [review]

**Comment by damienmiheev** [comment]
Maybe this is a good place to add a test for asynchronous enforcement?

**Comment by shijianjs** [comment]
@hsluoyz @imp2002 
After 30 seconds of running the test code #292 , there are other problems.

jcasbin version: 1.28.2、1.29.0
exception: 
```
java.lang.ClassCastException: class java.util.HashMap$Node cannot be cast to class java.util.HashMap$TreeNode (java.util.HashMap$Node and java.util.HashMap$TreeNode are in module java.base of loader 'bootstrap')

	at java.base/java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1882)
	at java.base/java.util.HashMap$TreeNode.treeify(HashMap.java:1998)
	at java.base/java.util.HashMap.treeifyBin(HashMap.java:767)
	at java.base/java.util.HashMap.putVal(HashMap.java:639)
	at java.base/java.util.HashMap.put(HashMap.java:607)
	at org.casbin.jcasbin.util.BuiltInFunctions$GenerateGFunctionClass$1.variadicCall(BuiltInFunctions.java:382)
	at com.googlecode.aviator.runtime.function.AbstractVariadicFunction.call(AbstractVariadicFunction.java:71)
	at Script_1663051805038_59.execute0(Script_1663051805038_59:1)
	at com.googlecode.aviator.ClassExpression.executeDirectly(ClassExpression.java:44)
	at com.googlecode.aviator.BaseExpression.execute(BaseExpression.java:237)
	at org.casbin.jcasbin.main.CoreEnforcer.enforce(CoreEnforcer.java:519)
	at org.casbin.jcasbin.main.CoreEnforcer.enforce(CoreEnforcer.java:631)
	at com.fingard.rdc.avatar.auth.casbin.CasbinTest.lambda$parallelTest$5(CasbinTest.java:97)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

```

**Pull request merged**